### PR TITLE
Optimize item migrations to decrease load times

### PR DESCRIPTION
**Pull request by aerrant**
#### Summary
Performance "Optimize item migrations to decrease load times"

#### Purpose of change
Presently there is a noticeable slowdown of one or two seconds on "Items" entry of "Finalizing" game load phase. It is caused by the excess of linear item id searches/replaces during item migrations processing and can be entirely eliminated, which I did.

#### Describe the solution
Instead of iterating over each migration individually, a map of migrations is constructed beforehand and migrations are applied in bulk. The `replace_item` methods were renamed to `replace_items` and updated to use the map.

#### Testing
The game uses migrations extensively, so failures during migrations will cause no end of error messages on game load.

#### Additional context
This fix is the simplest that I could come up with. The affected classes do need a bit of refactoring, though - it is the choice of backing data structures that led to the original problem to begin with. But I'm not comfortable enough to do refactorings yet, sorry.

### CODE REVIEW

File: src/item_group.cpp [hunk 1]
```diff
@@ -288,19 +289,21 @@ bool Single_item_creator::remove_item( const itype_id &itemid )
...
 {
     if( modifier ) {
-        modifier->replace_item( itemid, replacementid );
+        modifier->replace_items( replacements );
     }
     if( type == S_ITEM ) {
-        if( itemid.str() == id ) {
-            id = replacementid.str();
+        auto it = replacements.find( itype_id( id ) );
+        if( it != replacements.cend() ) {
```
Comment (jbytheway) [inline src/item_group.cpp:300]:
```suggestion
        if( it != replacements.end() ) {
```
Using `cend` is equivalent to `end`, and less idiomatic.

File: src/requirements.cpp [hunk 2]
```diff
@@ -965,39 +966,35 @@ void requirement_data::blacklist_item( const itype_id &id )
...
-                target = iter;
-            } else if( iter->type == replacement ) {
-                replacement_target = iter;
+static void apply_replacements( std::vector<std::vector<T>> &vec,
+                                const std::unordered_map<itype_id, itype_id> &replacements )
+{
+    for( std::vector<T> &opts : vec ) {
+        for( typename std::vector<T>::iterator iiter = opts.begin(); iiter != opts.end(); ) {
+            auto riter = replacements.find( iiter->type );
+            if( riter != replacements.cend() ) {
```
Comment (jbytheway) [inline src/requirements.cpp:975]:
```suggestion
            if( riter != replacements.end() ) {
```

File: src/requirements.cpp [hunk 3]
```diff
@@ -965,39 +966,35 @@ void requirement_data::blacklist_item( const itype_id &id )
...
+                itype_id to_id = riter->second;
+                // Replace an item or outright remove it if its replacement
+                // is already present in the vector, to prevent duplicates
+                bool has_duplicates = std::count_if( opts.cbegin(), opts.cend(), [&to_id]( const T & item ) {
+                    return item.type == to_id;
+                } ) != 0;
```
Comment (jbytheway) [inline src/requirements.cpp:981]: I would have used `find_if` here, which has the advantage of early exit.  On the other hand, `count_if` is less verbose, so I can see the argument for it.

**Review by jbytheway — COMMENTED** [review]
Looks generally reasonable, and a sensible algorithmic improvement.

**Comment by aerrant** [comment]
Thank you for your review, @jbytheway! I'll keep your begin()/end() suggestion in mind in the future.

Still, I'm not satisfied how this turned out: now that I've actually measured each entry of the game load process, I see that this PR brought the "Items" finalization phase from some 1700ms to 100ms, which is relatively good, but absolutely horrible, seeing how there are mere 265 migrations to be done on 5921 item templates and 3351 item groups to be processed. 3d games generally process much more data in less than a millisecond. I'll put the PR on ice to try to bring this down further to an arbitrary interval of 8ms (half a frame at 60fps) or less.

Here's the load time measurements I've made, for the reference (json parsing and entries of 8ms or less are omitted, and the margin of error is at some +-15%):

> Load, faster-migrations (-Os):
>     >>>>>
>     Finalized <Items> in <109> ms
>     Finalized <Overmap terrain> in <22> ms
>     Finalized <Vehicle prototypes> in <132> ms
>     Finalized <Mapgen weights> in <2132> ms
>     Finalized <Monster types> in <12> ms
>     Finalized <Crafting recipes> in <378> ms
>     Finalized <Tileset> in <780> ms
>     <<<<< finalization total <3985> ms
>     >>>>>
>     Verified <Items> in <910> ms
>     Verified <Mapgen definitions> in <273> ms
>     Verified <Overmap terrain> in <11> ms
>     <<<<< verification total <1916> ms

There's still much work to be done.

**Comment by jbytheway** [comment]
In this case I'd say you shouldn't let perfect be the enemy of good, but I know how fun it can be to optimise, so I won't stop you. Just bear in mind that if you make changes with small improvements to performance that make the code ltd clear, they probably won't get merged.

Also note that high performing games like you're describing go to great lengths to lay out their data in an efficient way. CDDA's data layout is absolutely awful, and that's unlikely to change.

**Comment by ZhilkinSerg** [comment]
Let's move it to post-0.F

**Comment by aerrant** [comment]
@ZhilkinSerg
I'm sorry for stalling. I think I won't be growing this PR's scope further - anything I can think of requires a thorough refactoring (e.g., merging the blacklist/migration loops) that'll need starting anew. For now, this PR seems low risk and useful - shouldn't it be fast tracked instead? The build failure isn't my fault for once, btw - it's the master branch that was red at the time :p

**Pull request merged**